flow/display_list_utils.cc

```diff
@@ -397,21 +396,71 @@ DisplayListBoundsCalculator::SaveLayerWithPaintInfo::SaveLayerWithPaintInfo(
     DisplayListBoundsCalculator* calculator,
     BoundsAccumulator* accumulator,
     const SkMatrix& saveMatrix,
+    const SkRect* saveBounds,
     const SkPaint& savePaint)
     : SaveLayerInfo(accumulator, saveMatrix),
       calculator_(calculator),
+      bounds_(saveBounds
+                  ? (saveBounds->isEmpty() ? SkRect::MakeEmpty() : *saveBounds)
+                  : kMissingBounds),
       paint_(savePaint) {}
 
+static bool PaintPreservesTransparency(const SkPaint& paint) {
```

Comment:
At some point Skia will make a public API for this function and we can delete this.

Reply:
Also, this is modeled after: https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/src/core/SkRecordDraw.cpp;drc=cfd204a78554f46651071f56ce518a83b8b3942b;l=312

but it handles more conditions and so provides a less conservative answer.